lang/shared/src/main/scala/com/wavesplatform/lang/v1/parser/Parser.scala

```diff
@@ -101,36 +101,54 @@ object Parser {
-    case Some(INVALID(_, endInvalid, nextXs, nextNext)) => foldInvalid(start, endInvalid, xs + nextXs, nextNext)
-    case x                                              => INVALID(start, end, xs, x)
-  }
-
   private val numberP: P[CONST_LONG] =
     P(Index ~~ (CharIn("+-").? ~~ digit.repX(min = 1)).! ~~ ("_" ~~ digit.repX(min = 1).!).repX(min = 0) ~~ Index)
       .map({ case (start, x1, x2, end) => CONST_LONG(start, end, x2.foldLeft(x1)(_ ++ _).toLong) })
 
   private val trueP: P[TRUE]        = P(Index ~~ "true".! ~~ Index).map { case (start, _, end) => TRUE(start, end) }
   private val falseP: P[FALSE]      = P(Index ~~ "false".! ~~ Index).map { case (start, _, end) => FALSE(start, end) }
-  private val bracesP: P[EXPR]      = P("(" ~ fallBackExpr ~ ")")
-  private val curlyBracesP: P[EXPR] = P("{" ~ fallBackExpr ~ "}")
-  private val letP: P[LET] = P(Index ~~ "let" ~ comment ~ anyVarName ~ comment ~ "=" ~/ fallBackExpr ~~ Index).map {
-    case (start, v, e, end) => LET(start, end, v, e, Seq.empty)
-  }
+  private val bracesP: P[EXPR]      = P("(" ~ baseExpr ~ ")")
+  private val curlyBracesP: P[EXPR] = P("{" ~ baseExpr ~ "}")
+
   private val refP: P[REF] = P(correctVarName).map { x =>
     REF(x.start, x.end, x)
   }
-  private val ifP: P[IF] = P(Index ~~ "if" ~ fallBackExpr ~ "then" ~ fallBackExpr ~ "else" ~ fallBackExpr ~~ Index).map {
-    case (start, x, y, z, end) => IF(start, end, x, y, z)
+
+  private val ifP: P[IF] = {
+    def optionalPart(keyword: String, branch: String): P[EXPR] = (Index ~ (keyword ~/ Index ~ baseExpr.?).?).map {
+      case (ifTruePos, ifTrueRaw) =>
+        ifTrueRaw
+          .map { case (pos, expr) => expr.getOrElse(INVALID(pos, pos, s"expected a $branch branch's expression")) }
+          .getOrElse(INVALID(ifTruePos, ifTruePos, s"expected a $branch branch"))
+    }
+
+    val thenPart = optionalPart("then", "true")
+    val elsePart = optionalPart("else", "false")
+
+    P(Index ~~ "if" ~/ Index ~ baseExpr.? ~ thenPart ~ elsePart ~~ Index).map {
+      case (start, condPos, condRaw, ifTrue, ifFalse, end) =>
+        val cond = condRaw.getOrElse(INVALID(condPos, condPos, "expected a condition"))
+        IF(start, end, cond, ifTrue, ifFalse)
+    } |
+      P(Index ~~ "then" ~/ Index ~ baseExpr.? ~ elsePart ~~ Index).map {
+        case (start, ifTrueExprPos, ifTrueRaw, ifFalse, end) =>
+          val ifTrue = ifTrueRaw.getOrElse(INVALID(ifTrueExprPos, ifTrueExprPos, "expected a true branch's expression"))
+          IF(start, end, INVALID(start, start, "expected a condition"), ifTrue, ifFalse)
+      } |
+      P(Index ~~ "else" ~/ Index ~ baseExpr.? ~~ Index).map {
+        case (start, ifFalseExprPos, ifFalseRaw, end) =>
+          val ifFalse = ifFalseRaw.getOrElse(INVALID(ifFalseExprPos, ifFalseExprPos, "expected a false branch's expression"))
+          IF(start, end, INVALID(start, start, "expected a condition"), INVALID(start, start, "expected a true branch"), ifFalse)
+      }
```

Comment:
Sad, but we can't simplify this expression, because we need a greedy parsing for incomplete IF expressions